src/test/java/eu/dissco/core/digitalspecimenprocessor/repository/DigitalSpecimenRepositoryIT.java

```diff
@@ -105,21 +101,21 @@ void testUpdateLastChecked() {
   }
 
   @Test
-  void testInsertDuplicateSpecimens() {
+  void testUpsertSpecimens() {
     // Given
     var records = List.of(
         givenDigitalSpecimenRecord(),
         givenDigitalSpecimenRecord(SECOND_HANDLE, "TEST_1"),
         givenDigitalSpecimenRecord(SECOND_HANDLE, "TEST_2"));
 
     // When
-    var exception = assertThrows(DataAccessException.class, () -> {
-      repository.createDigitalSpecimenRecord(records);
-    });
+    repository.createDigitalSpecimenRecord(records);
 
     // Then
-    assertThat(exception).hasCauseInstanceOf(BatchUpdateException.class).hasRootCauseInstanceOf(
-        PSQLException.class);
+    var result = context.select(NEW_DIGITAL_SPECIMEN.PHYSICAL_SPECIMEN_ID)
+        .from(NEW_DIGITAL_SPECIMEN).where(NEW_DIGITAL_SPECIMEN.ID.eq(SECOND_HANDLE))
+        .fetchOne(Record1::value1);
+    assertThat(result).isEqualTo("TEST_2");
   }
```

Comment:
So, did this function change to no longer insert duplicate specimens with a new version? Instead create assert a new instance? Can you elaborate a bit on this?

Reply:
Yes ofcourse. So previously we created a new record for each update. So you had record version 1 then an update came in and we create record version 2. So now we have two records, version 1 and version 2. This has now been changed to an upsert (insert when there is nothing otherwise update). Now when version 1 comes we created the record. When version 2 comes in we update the first record with the new data. So now we have 1 record with version 2. So instead of creating a new record we updated the existing one with the new information.

This test first checked if we did not insert the same specimen / version into the database, if so it would throw an error. However now if we get new information we just update the database instead of fail as updating records is now the required behaviour.

Reply:
That explains :+1: